api/proto/teleport/legacy/types/types.proto

```diff
@@ -1892,14 +1897,155 @@ message ProvisionTokenSpecV2Oracle {
+    // `initial_public_key` is provided, a random secure value will be generated
+    // server-side to populate the status field.
+    string InitialJoinSecret = 2 [(gogoproto.jsontag) = "initial_join_secret,omitempty"];
+
+    // MustJoinBefore is an optional time before which initial secret joining
+    // may be used. Attempts to register using an initial join secret after this
+    // timestamp will not be allowed. This may be modified after creation if
+    // necessary to allow the initial registration to take place.
+    google.protobuf.Timestamp MustJoinBefore = 3 [
+      (gogoproto.stdtime) = true,
+      (gogoproto.nullable) = true,
+      (gogoproto.jsontag) = "must_join_before,omitempty"
+    ];
+  }
+
+  message JoiningSpec {
+    // Unlimited indicates no limit for joining and rejoining. `total_joins`
+    // and the calculated `.status.bound_keypair.join_count` value will
+    // increment as usual, but join attempts will succeed even if the
+    // `total_joins` limit is exceeded.
+    bool Unlimited = 1 [(gogoproto.jsontag) = "unlimited"];
+
+    // TotalJoins is the total number of joins that will be allowed over the
+    // lifetime of this token. This value may be raised or lowered after
+    // creation to allow additional joins should
+    // `.status.bound_keypair.join_count` reach this limit. Unless `unlimited`
+    // is set, this value must be greater than `join_count` for a join attempt
+    // to succeed. The initial join counts toward this limit, so it must be at
+    // least 1 to allow onboarding. If `unlimited` is set, this value can be set
+    // to 0 and ignored.
+    uint32 TotalJoins = 2 [(gogoproto.jsontag) = "total_joins"];
+
+    // MayJoinUntil is an optional timestamp before which all joining attempts
+    // must occur. If unset, there are no time constraints on joining attempts.
+    // Note that clients may continue to renew their identities past this
+    // timestamp; a lock should be created if continued access should be cut
+    // off. This value may be modified after creation to adjust the joining
+    // deadline as desired.
+    google.protobuf.Timestamp MayJoinUntil = 3 [
+      (gogoproto.stdtime) = true,
+      (gogoproto.nullable) = true,
+      (gogoproto.jsontag) = "may_join_until"
+    ];
+
+    // Insecure is an optional flag that enables insecure joining and rejoining.
+    // When set, join state verification and generation counter checks are
+    // disabled, allowing an arbitrary number of "initial" joins. When combined
+    // with the `unlimited` flag, this allows unlimited reuse of this token
+    // provided the client has access to the keypair. This may be useful in
+    // certain cases - like use in unsupported CI/CD providers - but cannot
+    // offer the same security assurances. This flag should be used with care,
+    // and RBAC rules should heavily restrict which resources this identity can
+    // access.
+    bool Insecure = 4 [(gogoproto.jsontag) = "insecure"];
```

Comment:
Another discussion note: we decided to merge Unlimited and Insecure into an enum field, something conceptually like:

```proto
enum RecoveryMode {
  STANDARD,
  UNLIMITED,
  UNLIMITED_INSECURE
}

RecoveryMode mode = 4;
```

(names subject to further change for clarity; if I can come up with good, clear mode names, I'll opt for that instead)

I'll also need to gauge the YAML (and hcl/k8s/etc) representations here. I think in practice this will be a server-validated string value since enum encoding is ... bad.

Reply:
I've tentatively gone with a string field with these possible values:
- `standard` - best security, recovery count is limited, and client state is verified
- `relaxed` - recovery count is not verified, client state is verified
- `insecure` - neither recovery count nor client state is verified

What do you think of these values and definitions?

Reply:
Perfect!